File: kotlin-labs/src/main/kotlin/javalr6/DrivingLicense.java
package javalr6;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.Objects;
import java.util.function.BiFunction;
import java.util.function.Function;

public class DrivingLicense implements Comparable<DrivingLicense> {
    private static final String SERIES_PATTERN = "^\\d{4}$";
    private static final String NUMBER_PATTERN = "^\\d{6}$";

    private final String series;
    private final String number;
    private final LocalDate issueDate;
    private final LocalDate expirationDate;

    public DrivingLicense(String series, String number, LocalDate issueDate, LocalDate expirationDate) {
        this.series = validateSeries(series);
        this.number = validateNumber(number);
        this.issueDate = issueDate;
        this.expirationDate = expirationDate;
    }

    @Override
    public String toString() {
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd.MM.yyyy");
        return "DrivingLicense(series='" + series + "', number='" + number + "', issueDate=" + issueDate.format(formatter) + ", expirationDate=" + expirationDate.format(formatter) + ")";
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        DrivingLicense that = (DrivingLicense) o;
        return series.equals(that.series) && number.equals(that.number);
    }

    @Override
    public int hashCode() {
        return Objects.hash(series, number);
    }

    @Override
    public int compareTo(DrivingLicense other) {
        return compareValuesBy(this, other,
                DrivingLicense::getSeries,
                DrivingLicense::getNumber,
                DrivingLicense::getIssueDate,
                DrivingLicense::getExpirationDate);
    }

    private String validateSeries(String series) {
        if (!series.matches(SERIES_PATTERN)) {
            throw new IllegalArgumentException("Неверная серия документа: " + series);
        }
        return series;
    }

    private String validateNumber(String number) {
        if (!number.matches(NUMBER_PATTERN)) {
            throw new IllegalArgumentException("Неверный номер документа: " + number);
        }
        return number;
    }

    public String getSeries() {
        return series;
    }

    public String getNumber() {
        return number;
    }

    public LocalDate getIssueDate() {
        return issueDate;
    }

    public LocalDate getExpirationDate() {
        return expirationDate;
    }

    private static <T> int compareValuesBy(T obj1, T obj2, Function<T, ?>... getters) {
        return compareValuesBy(obj1, obj2, 0, getters);
    }

    private static <T> int compareValuesBy(T obj1, T obj2, int index, Function<T, ?>[] getters) {
        if (index >= getters.length) {
            return 0;
        }

        Function<T, ?> getter = getters[index];
        Object val1 = getter.apply(obj1);
        Object val2 = getter.apply(obj2);

        int comparison = compareValues(val1, val2);
        if (comparison != 0) {
            return comparison;
        }

        return compareValuesBy(obj1, obj2, index + 1, getters);
    }

    private static int compareValues(Object val1, Object val2) {
        if (val1 == null && val2 == null) {
            return 0;
        } else if (val1 == null) {
            return -1;
        } else if (val2 == null) {
            return 1;
        } else if (val1 instanceof Comparable<?> && val2 instanceof Comparable<?>) {
            Comparable c1 = (Comparable<?>) val1;
            Comparable c2 = (Comparable<?>) val2;
            return c1.compareTo(c2);
        } else {
            return val1.toString().compareTo(val2.toString());
        }
    }
}

File: kotlin-labs/src/main/kotlin/LR6/DrivingLicense.kt
package LR6

import java.time.LocalDate
import java.time.format.DateTimeFormatter

class DrivingLicense(
    val series: String,
    val number: String,
    val issueDate: LocalDate,
    val expirationDate: LocalDate,
) : Comparable<DrivingLicense> {

    companion object {
        private val seriesPattern = Regex("^\\d{4}$")
        private val numberPattern = Regex("^\\d{6}$")
    }

    init {
        validateSeries(series)
        validateNumber(number)
    }

    override fun toString(): String {
        val formatter = DateTimeFormatter.ofPattern("dd.MM.yyyy")
        return "DrivingLicense(series='$series', number='$number', issueDate=${issueDate.format(formatter)}, expirationDate=${
            expirationDate.format(
                formatter
            )
        })"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as DrivingLicense

        if (series != other.series) return false
        if (number != other.number) return false

        return true
    }

    override fun hashCode(): Int {
        var result = series.hashCode()
        result = 31 * result + number.hashCode()
        return result
    }

    override fun compareTo(other: DrivingLicense): Int {
        val seriesNumberComparison = compareValuesBy(this, other, { it.series }, { it.number })
        if (seriesNumberComparison != 0) return seriesNumberComparison

        val issueDateComparison = issueDate.compareTo(other.issueDate)
        if (issueDateComparison != 0) return issueDateComparison

        return expirationDate.compareTo(other.expirationDate)
    }

    private fun validateSeries(series: String) {
        if (!seriesPattern.matches(series)) {
            throw IllegalArgumentException("Неверная серия документа: $series")
        }
    }

    private fun validateNumber(number: String) {
        if (!numberPattern.matches(number)) {
            throw IllegalArgumentException("Неверный номер документа: $number")
        }
    }
}